pkg/blockdevice/v1alpha2/kubernetes.go

```diff
@@ -0,0 +1,173 @@
+// getFn is a typed function that
+// abstracts fetching of block deivce
+type getFn func(cli *clientset.Clientset, namespace, name string, opts metav1.GetOptions) (*apis.BlockDevice, error)
+
+// Kubeclient enables kubernetes API operations
+// on block device instance
+type Kubeclient struct {
+	// clientset refers to block device
+	// clientset that will be responsible to
+	// make kubernetes API calls
+	clientset *clientset.Clientset
+	// kubeconfig path to get kubernetes clientset
+	kubeConfigPath string
+	namespace      string
+	// functions useful during mocking
+	getClientset        getClientsetFn
+	getClientsetForPath getClientsetForPathFn
+	list                listFn
+	get                 getFn
+}
+
+// KubeclientBuildOption defines the abstraction
+// to build a kubeclient instance
+type KubeclientBuildOption func(*Kubeclient)
+
+// WithDefaults sets the default options
+// of kubeclient instance
+func (k *Kubeclient) WithDefaults() {
+	if k.getClientset == nil {
+		k.getClientset = func() (clients *clientset.Clientset, err error) {
+			config, err := kclient.New().Config()
+			if err != nil {
+				return nil, err
+			}
+			return clientset.NewForConfig(config)
+		}
+	}
+	if k.getClientsetForPath == nil {
+		k.getClientsetForPath = func(kubeConfigPath string) (clients *clientset.Clientset, err error) {
+			config, err := kclient.New(kclient.WithKubeConfigPath(kubeConfigPath)).Config()
+			if err != nil {
+				return nil, err
+			}
+			return clientset.NewForConfig(config)
+		}
+	}
+	if k.list == nil {
+		k.list = func(cli *clientset.Clientset, namespace string, opts metav1.ListOptions) (*apis.BlockDeviceList, error) {
+			return cli.OpenebsV1alpha1().BlockDevices(namespace).List(opts)
+		}
+	}
+
+	if k.get == nil {
+		k.get = func(cli *clientset.Clientset, namespace, name string, opts metav1.GetOptions) (*apis.BlockDevice, error) {
+			return cli.OpenebsV1alpha1().BlockDevices(namespace).Get(name, opts)
+		}
+	}
+}
+
+// WithKubeClient sets the kubernetes client against
+// the kubeclient instance
+func WithKubeClient(c *clientset.Clientset) KubeclientBuildOption {
+	return func(k *Kubeclient) {
+		k.clientset = c
+	}
+}
+
+// WithKubeConfigPath sets the kubeConfig path
+// against client instance
+func WithKubeConfigPath(kubeConfigPath string) KubeclientBuildOption {
+	return func(k *Kubeclient) {
+		k.kubeConfigPath = kubeConfigPath
+	}
+}
+
+// NewKubeClient returns a new instance of kubeclient meant for
+// block device operations
+func NewKubeClient(opts ...KubeclientBuildOption) *Kubeclient {
+	k := &Kubeclient{}
+	for _, o := range opts {
+		o(k)
+	}
+	k.WithDefaults()
+	return k
+}
+
+func (k *Kubeclient) getClientsetForPathOrDirect() (*clientset.Clientset, error) {
+	if k.kubeConfigPath != "" {
+		return k.getClientsetForPath(k.kubeConfigPath)
+	}
+	return k.getClientset()
+}
+
+// WithNamespace sets the kubernetes namespace against
+// the provided namespace
+func (k *Kubeclient) WithNamespace(namespace string) *Kubeclient {
```

Comment:
can this be also added as `KubeclientBuildOption`

Reply:
There will be a cases where we need to handle `blockdeviceclaim` present in different namespaces if we have as a `KubeclientBuildOption` then for different namespace we need to get clientset for `blockdeviceclaim` again and again.

Reply:
@vishnuitta both ways should be supported